chia/wallet/wallet_state_manager.py

```diff
@@ -855,12 +855,9 @@ async def handle_cat(
         """
         mod_hash, tail_hash, inner_puzzle = curried_args
 
-        hint_dict, _ = compute_spend_hints_and_additions(coin_spend)
-        coin_name: bytes32 = bytes32(coin_state.coin.name())
-        if coin_name in hint_dict:
-            derivation_record = await self.puzzle_store.get_derivation_record_for_puzzle_hash(hint_dict[coin_name])
-        else:
-            derivation_record = None
+        hinted_coin = compute_spend_hints_and_additions(coin_spend)[coin_state.coin.name()]
+        assert hinted_coin.hint is not None, f"hint missing for coin {hinted_coin.coin}"
```

Comment:
Sorry, one last comment here, I believe I've commented this on some PR at some point, but I think raising here is not advisable because we might try to sync some unhinted change or something and this is a perfectly normal failure and we want to handle it gracefully rather than flood a user's logs with this as a warning.

Reply:
The DID one below might be fine, but for the sake of consistency maybe it's better to fix that one too.

Reply:
> this is a perfectly normal failure

Why is this perfectly normal? From my understanding not hinting CATs is and will ever be an issue and if we again due to some bug somewhere start to not hint CATs or some other wallet misses it we at least will be alerted about it this way.

And why do you think its better/less confusing for users to log a bunch of `Received state for the coin that doesn't belong to us` messages (which happens if we go your way) which basically is not true here instead of logging the actual issue?

Reply:
Which brings me to the question, how do we even sync unhinted CATs? Yes, we have a test which subscribes to them so that we at least get them synced but if there is no hint we can not figure out the derivation record here anyway and we just step out of it and assume it doesn't belong to us, no?

Reply:
Yeah my bad, I didn't have it fully swapped into my brain, but you're right we pick up change via the puzhash in the DB, we shouldn't be going through this function.

I do still feel that this should not be a raise because you're basically short circuiting any potential logic for "we couldn't determine the coin type here".  Looking at the existing logic though, this seems fine currently, I just think it's a bad choice going forward were someone to add something that we are unaware of currently (maybe a like "we don't know what this is" store or something).

Since it works in the existing code though and I'd really just like to get this merged, you can resolve if you disagree.

Reply:
> we pick up change via the puzhash in the DB, we shouldn't be going through this function.

Oh right so this works because we still derive/store the puzzle hashes and then we get the wallet identifier from looking up the puzzle store. But we still run into `handle_cat` if the CAT wasn't added yet? So wouldn't syncing change also fail if we sync the change coin as first coin of the given CAT, i.e. out of order?

> I do still feel that this should not be a raise because you're basically short circuiting any potential logic for "we couldn't determine the coin type here".

Not sure i get what you mean here.

> And why do you think its better/less confusing for users to log a bunch of Received state for the coin that doesn't belong to us messages (which happens if we go your way) which basically is not true here instead of logging the actual issue?

Did you read my comment ^ about logging the actual issue vs logging "this doesn't belong to us" what you propose?

Reply:
> Oh right so this works because we still derive/store the puzzle hashes and then we get the wallet identifier from looking up the puzzle store. But we still run into handle_cat if the CAT wasn't added yet? So wouldn't syncing change also fail if we sync the change coin as first coin of the given CAT, i.e. out of order?

Right, we need at least one hinted CAT, but the thing is that it's actually impossible for change to come out of order now because the only way we pick it up is when the parent gets spent.  Even if it did come out of order, I think the new subscription when the parent gets spent will pick it up again (it's a coin ID subscription).

> Not sure i get what you mean here.

Right now, we log something then return None from the `determine_coin_type` function and then the loop continues.  Right now almost nothing happens and we just continue to the next state, but it's possible that in the future someone might want to do something different and they'll have to undo your change here in order to do so.

> Did you read my comment ^ about logging the actual issue vs logging "this doesn't belong to us" what you propose?
Well if it's not change like we've identified, then that is the correct error message.  "I got this state, but I can't verify it belongs to me (because it wasn't hinted and I don't know what the inner hash is)".

Reply:
> Right now, we log something then return None from the `determine_coin_type` function and then the loop continues. Right now almost nothing happens and we just continue to the next state

Now we properly log an error, which is what we should do in a case a not hinted CAT comes in imo, and then we continue with the loop just like we would in your case.

The difference is in your case we would log something which isn't helpful in any way and silently step over it. It doesn't allow to distinguish between unhinted CATs and CATs which really don't belong to our wallet, it would both look the same and we wouldn't become noticed about coin states we can't process because of this issue but instead it would like they just don't belong to us.

> but it's possible that in the future someone might want to do something different and they'll have to undo your change here in order to do so.

Maybe i miss something here but isn't the only thing which would make sense to do different in this case trying to fetch the spend of this coin and see if we can get the puzzle to identify the CAT which should happen in `handle_cat` anyway? 

> I got this state, but I can't verify it belongs to me (because it wasn't hinted and I don't know what the inner hash is)"

Thats not what the log would say though. I just logs an `INFO` with `Received state for the coin that doesn't belong to us`.

Reply:
Btw, the failing CI is, i think, only because your branch doesn't contain the mypy 1.4 fixes which recently came up so if you merge it in it should be fine because your branch goes into `main`.